samples/snippets/csharp/VS_Snippets_CLR/conceptual.disposable/cs/base1.cs

```diff
@@ -6,84 +6,84 @@
 
-   // Define Windows APIs.
-   [DllImport("kernel32.dll", EntryPoint = "CreateFileW", CharSet = CharSet.Unicode)]
-   protected static extern IntPtr CreateFile (
-                                  string lpFileName, uint dwDesiredAccess,
-                                  uint dwShareMode, IntPtr lpSecurityAttributes,
-                                  uint dwCreationDisposition, uint dwFlagsAndAttributes,
-                                  IntPtr hTemplateFile);
+    // Define Windows APIs.
+    [DllImport("kernel32.dll", EntryPoint = "CreateFileW", CharSet = CharSet.Unicode)]
+    protected static extern IntPtr CreateFile(
+        string lpFileName, uint dwDesiredAccess,
+        uint dwShareMode, IntPtr lpSecurityAttributes,
+        uint dwCreationDisposition, uint dwFlagsAndAttributes,
+        IntPtr hTemplateFile);
 
-   [DllImport("kernel32.dll")]
-   private static extern int GetFileSize(SafeFileHandle hFile, out int lpFileSizeHigh);
+    [DllImport("kernel32.dll")]
+    static extern int GetFileSize(SafeFileHandle hFile, out int lpFileSizeHigh);
 
-   // Define locals.
-   private bool disposed = false;
-   private SafeFileHandle safeHandle;
-   private long bufferSize;
-   private int upperWord;
+    // Define locals.
+    bool _disposed = false;
+    readonly SafeFileHandle _safeHandle;
+    readonly int _upperWord;
 
-   public DisposableStreamResource(string filename)
-   {
-      if (filename == null)
-         throw new ArgumentNullException("The filename cannot be null.");
-      else if (filename == "")
-         throw new ArgumentException("The filename cannot be an empty string.");
+    public DisposableStreamResource(string fileName)
+    {
+        if (string.IsNullOrWhiteSpace(fileName))
+        {
+            throw new ArgumentException("The fileName cannot be null or an empty string");
+        }
 
-      IntPtr handle = CreateFile(filename, GENERIC_READ, FILE_SHARE_READ,
-                                 IntPtr.Zero, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL,
-                                 IntPtr.Zero);
-      if (handle != INVALID_HANDLE_VALUE)
-         safeHandle = new SafeFileHandle(handle, true);
-      else
-         throw new FileNotFoundException(String.Format("Cannot open '{0}'", filename));
+        IntPtr handle = CreateFile(
+            fileName, GENERIC_READ, FILE_SHARE_READ, IntPtr.Zero,
+            OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, IntPtr.Zero);
 
-      // Get file size.
-      bufferSize = GetFileSize(safeHandle, out upperWord);
-      if (bufferSize == INVALID_FILE_SIZE)
-         bufferSize = -1;
-      else if (upperWord > 0)
-         bufferSize = (((long)upperWord) << 32) + bufferSize;
-   }
+        if (handle == INVALID_HANDLE_VALUE)
+        {
+            throw new FileNotFoundException($"Cannot open '{fileName}'");
+        }
 
-   public long Size
-   { get { return bufferSize; } }
+        _safeHandle = new SafeFileHandle(handle, true);
```

Comment:
We shouldn't be advocating this style; it's necessary sometimes, but it's the 1% case.  The runtime's marshaling infrastructure is capable of doing this for the developer, so that they don't need to deal with the underlying IntPtr at all (it also helps with reliability on platforms that have thread abort).  Instead of declaring the P/Invoke to return an IntPtr, you just declare it to return the SafeHandle type, and the runtime itself is then responsible for creating the SafeHandle around the returned IntPtr.